Проект 1 – ДНК

evgeny-golubev opened the pull request.

evgeny-golubev committed:
v 0.3

evgeny-golubev committed:
v 0.4

evgeny-golubev committed:
Revert "v 0.4"

This reverts commit 8b024a3b67a96b0215a9485a2cf0efe1287ff2d2.

evgeny-golubev committed:
Revert "v 0.3"

This reverts commit fcd2162f02c5f38c86e2eaf1640a367ea121ded0.

evgeny-golubev committed:
Revert "Revert "v 0.4""

This reverts commit d8346de372b2a44efdbab5fa2f48850b9fce969b.

evgeny-golubev committed:
Revert "Revert "v 0.3""

This reverts commit 7c3e0840670f91be133fa7bc5595af3da5a04250.

evgeny-golubev committed:
Revert "v 0.4"

This reverts commit 8b024a3b67a96b0215a9485a2cf0efe1287ff2d2.

evgeny-golubev committed:
Revert "Revert "v 0.4""

This reverts commit e09c1733f4e00f9e2bb50e6e802a8603cb102a39.

evgeny-golubev committed:
Revert "Revert "v 0.4""

This reverts commit d8346de372b2a44efdbab5fa2f48850b9fce969b.

evgeny-golubev committed:
Revert "Revert "v 0.3""

This reverts commit 7c3e0840670f91be133fa7bc5595af3da5a04250.

evgeny-golubev committed:
Revert "Revert "Revert "v 0.3"""

This reverts commit 8d8347f78e9b302116a2d34d80cc373c5678606d.

evgeny-golubev committed:
Revert "Revert "Revert "v 0.4"""

This reverts commit 1b508aa4a1d435b08d3ad976588dafe66a6a8b96.

evgeny-golubev committed:
v 0.4

Update

evgeny-golubev committed:
v 1.0

Итоговый билд для отправки на проверку.

evgeny-golubev committed:
v 1.1

evgeny-golubev committed:
v 1.1

Обновление

evgeny-golubev committed:
добавлены игнорирования

evgeny-golubev committed:
Обновление

evgeny-golubev committed:
Обновление

cleg commented on DNA Test/DNA Test/Cell+Mutator.m:
```diff
@@ -0,0 +1,41 @@
+	//Проверяем процент мутации и запускаем мутацию только в случае его корректности
+    if (mutatePercent>0 && mutatePercent<=100) {
+		//Определяем колличество элементов для мутации в случае если длинна ДНК больше 100
+        int mutateIndexesCount = DNA_LENGHT*mutatePercent/100;
```
Тут лучше использовать вещественные числа и потом округление

evgeny-golubev replied:
А можно подробнее?

cleg replied:
Ну вот допустим у нас 90 элементов в массиве ДНК. И нас просят заменить 65% 
Если посчитать в вещественных числах — получается 90*65/100 = 58,5, округлив — получаем что надо заменить 59 элементов
В вашем же случае — целочисельная арифметика даст 58.

Я просто сделал вот так:
NSUInteger needToMutate = lround(percent/100.0 \* DNA_SIZE);

evgeny-golubev replied:
Логично, спасибо.

cleg commented on DNA Test/DNA Test/Cell+Mutator.m:
```diff
@@ -0,0 +1,41 @@
+		//Для получения уникального набора индексов создаём изменяемое множество
+        NSMutableSet *mutateIndexes = [[NSMutableSet alloc] initWithCapacity:mutateIndexesCount];
+		
+		//Пока множество не заполниться генерируем туда случайный индекс по длинне ДНК, особенность уникальности элеменов в множестве не даст дублировать индекс
```
Достаточно неоптимальный метод, так как на процентах близких к 100 — будет требовать много лишних итераций.
В данном конкретом случае это не очень критично, но на больших числах это уже может быть проблемой.

cleg commented on DNA Test/DNA Test/Cell+Mutator.m:
```diff
@@ -0,0 +1,41 @@
+		
+		//Проходим по всем элементам в множестве
+        for (id mutateIndex in mutateIndexes) {
+			//Создаём изменяемый массив на основе наших нуклиотидов
```
Честно говоря — дальше не понял код. Зачем копировать массивы, если можно просто генерировать новые случайные "буквы", и если не совпадает с тем что в этой позиции — заменять

cleg commented on DNA Test/DNA Test/Cell.m:
```diff
@@ -0,0 +1,52 @@
+
+    if (self) {
+        //Инициализируем ДНК необходимой длинны
+        DNA = [[NSMutableArray alloc] initWithCapacity:DNA_LENGHT];
```
Удобный шорткат в данном случае:
NSMutableArray *DNA = [NSMutableArray arrayWithCapacity:DNA_SIZE];

cleg commented on DNA Test/DNA Test/Cell.m:
```diff
@@ -0,0 +1,52 @@
+
+-(NSArray*) getNucleotides {
+	//Возвращаем массив с нуклотидами
+    return [NSArray arrayWithObjects:@"A",@"T",@"G",@"C",nil];
```
У вас при каждом вызове этой ф-ии конструируется массив и в него помещаются инстансы NSString — а это весьма накладно. 
Было бы логичней сделать массив статическим, первый раз инициализировать, и потом возвращать.

Плюс — я бы сделал метод, возвращающий не массив, а сразу случайный его элемент. Так как сам по себе этот массив вроде бы не нужен нигде — из него сразу же выбирается случайный элемент.

evgeny-golubev commented:
Спасибо, буду дорабатывать. Это пока не Final :) Ваши замечания учту.

evgeny-golubev committed:
v 1.2

Обновления мутации. Нуклеотиды вынесены в статический массив. Добавлен
метод класса для получения случайного нуклеоида.

evgeny-golubev committed:
Final

No changes

mitskevich commented on DNA Test/DNA Test/Cell.m:
```diff
@@ -0,0 +1,61 @@
+    //Инициализируем счётчик несоответствий
+    int hamming = 0;
+    
+    for (int i=1; i<DNA_LENGHT; i++) {
```
Эта структура пропустит первую ячейку массива (индексы начинаются с 0).

mitskevich commented on DNA Test/DNA Test/Cell.m:
```diff
@@ -0,0 +1,61 @@
+    
+    for (int i=1; i<DNA_LENGHT; i++) {
+        //Сравниваем поэлементно и в случае несоответствия увеличиваем счётчик
+        if([[DNA objectAtIndex:i] isEqual:[[DNAForHamming getDNA] objectAtIndex:i]]) {
```
По условию, hammingDistance должен быть равен числу несовпадающих ячеек. Эта реализация выдаёт наоборот число совпадающих (проверка на совпадение).

evgeny-golubev committed:
Исправление по комментариям.

evgeny-golubev committed:
one more thing...
